design/re-introspection-interface.md

```diff
@@ -0,0 +1,90 @@
+# re-introspection API
+
+## Status
+
+implementable
+
+## Summary
+
+A declarative API is proposed to request the baremetal operator to
+re-introspect a `Ready` BareMetalHost.
+
+## Motivation
+
+We would like to have an interface to allow a user to re-gather hardware
+inventory of a `Ready` BareMetalHost when a hardware replacement is made.
+When a user of the underlying infrastructure makes some changes to the actual
+server (e.g. replace or add NIC, disk, etc.), the latest hardware inventory
+including those changes need to be re-collected and updated on the spec of the
+corresponding BareMetalHost object without having to delete it.
+
+Implementation of this proposal is based on using annotation (similar to
+[Reboot API](https://github.com/metal3-io/metal3-docs/blob/master/design/reboot-interface.md))
+to request re-introspection of a `Ready` BareMetalHost.
+Once the annotation is set on BareMetalHost, the baremetal operator will
+request hardware introspection of the host from
+Ironic.
+
+### Goals
+
+- A declarative API to perform re-introspection
+- Use this API for future Metal³ remediation controller
+
+### Non-Goals
+
+- Automated mechanism to trigger re-introspection
+
+## Proposal
+
+We follow Reboot API implementation and expect to implement the similar
+annotation based interface for a user to trigger re-introspection.
+
+`reintrospection.metal3.io` annotation form on BareMetalHost object
+triggers the controller to query Ironic for introspection of a host. This form
+has set-and-forget semantics and controller removes the annotation once
+introspection is completed.
+While the host is being re-introspected, the BareMetalHost will stay in
+`Inspecting` state until the process is completed.
+
+Re-introspection API can be requested only when BareMetalHost is in `Ready`
+state. If a re-introspection request is made while BareMetalHost is any other
+state than `Ready`, the request will be ignored, i.e., annotation will
+not be set on BareMetalHost object. This is important in order to not reboot a
+BareMetalHost (e.g. when `spec.provisioning.state == provisioned`) or avoid
+having unintended inconsistent states.
+
+|BMH state|Externally provisioned|API action|Annotation|
+|---|---|---|---|
+|Ready|False|call introspection|delete|
```

Comment:
Rather than "call introspection" I would say "move to Inspecting state".

It may not be possible to delete the annotation as part of that state change; I'm not sure it's possible to atomically update both status and annotations at the same time. In any event, there's no need for this. Just delete the annotation in the Inspecting state.

Reply:
Updated.